torch/export/unflatten.py

```diff
@@ -711,62 +709,74 @@ def unflatten(
     return UnflattenedModule(module, flat_args_adapter)
 
 
-def _inplace_buffer_mutations(
+def _inplace_buffer_and_input_mutations(
     graph: torch.fx.Graph,
     graph_signature: ExportGraphSignature,
 ) -> None:
-    """Transform buffer mutations from their functionalized form into a copy_
-    node in the graph.
+    """Transform buffer and input mutations from their functionalized form
+    into copy_ nodes in the graph.
 
-    Functionalization represents buffer mutation by passing the buffer as an input and output. So for example, the eager code:
+    Functionalization represents a buffer mutation by passing the buffer as
+    an input and output. For example, consider the eager code:
         def forward(self, x):
             self.buffer += x
             return x * x
 
-    Will become a graph that looks like:
+    This corresponds to a graph that looks like:
         def forward(self, buffer, x):
             mutated_buffer = aten.add(buffer, x)
             mul = aten.mul(x, x)
             return (mutated_buffer, mul)
 
-    We want to inplace this into something that looks like the original eager code:
+    We want to inplace this into something that looks like the original
+    eager code:
         def forward(self, buffer, x):
             mutated_buffer = aten.add(buffer, x)
             buffer.copy_(mutated_buffer)
             mul = aten.mul(x, x)
             return (mul,)
+
+    Input mutations are handled similarly.
     """
     output_node = next(iter(reversed(graph.nodes)))
     assert output_node.op == "output" and len(output_node.args) == 1
     return_args = output_node.args[0]
 
-    mutation_node_to_buffer = graph_signature.buffers_to_mutate
-    mutations = return_args[: len(mutation_node_to_buffer)]
-    buffers_to_inputs = {v: k for k, v in graph_signature.inputs_to_buffers.items()}
     input_name_to_node = {
         node.name: node for node in graph.nodes if node.op == "placeholder"
     }
+    mutation_name_to_input_name = {}
+
+    # Collect mutated buffers.
+    buffer_fqn_to_input_name = {
+        buffer_fqn: k for k, buffer_fqn in graph_signature.inputs_to_buffers.items()
+    }
+    for k, buffer_fqn in graph_signature.buffers_to_mutate.items():
```

Comment:
In training IR, this signature is empty.

Reply:
Or i guess in training IR, we would have returned earlier?

Reply:
no problem with training IR because the mutations are already in the graph